File: src/main/java/net/hotamachisubaru/digger/mysql/MySQLDatabase.java
package net.hotamachisubaru.digger.mysql;
import net.hotamachisubaru.digger.Digger;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import java.io.FileInputStream;
import java.io.IOException;
import java.sql.*;
import java.util.*;

public class MySQLDatabase {
    private String url;
    private String user;
    private String password;

    public MySQLDatabase(Properties prop) {
        try {
            prop.load(new FileInputStream(Digger.getInstance().getDataFolder().getAbsolutePath() + "/config.properties"));
            this.url = prop.getProperty("db.url");
            this.user = prop.getProperty("db.user");
            this.password = prop.getProperty("db.password");
            initializeDatabase();
        } catch (IOException e) {
            Bukkit.getLogger().severe("MySQLデータベースに接続できませんでした。");
            // エラーハンドリング...
        }
    }

    private void initializeDatabase() {
        try (Connection conn = getConnection();
             Statement stmt = conn.createStatement()) {
            // player_data テーブルの作成
            String playerDataTable = "CREATE TABLE IF NOT EXISTS digger_player_data ("
                    + "UUID CHAR(36) PRIMARY KEY,"
                    + "PlayerName VARCHAR(255),"
                    + "BlocksMined INT);";
            stmt.execute(playerDataTable);

            // placed_blocks テーブルの作成
            String placedBlocksTable = "CREATE TABLE IF NOT EXISTS digger_placed_blocks ("
                    + "UUID CHAR(36),"
                    + "World VARCHAR(255),"
                    + "X INT,"
                    + "Y INT,"
                    + "Z INT,"
                    + "FOREIGN KEY(UUID) REFERENCES digger_player_data(UUID));";
            stmt.execute(placedBlocksTable);
        } catch (SQLException e) {
            Bukkit.getLogger().severe("MySQLデータベースに保存できませんでした。接続を確認してください。");
            // エラーハンドリング...
        }
    }

    private Connection getConnection() throws SQLException {
        // MySQLデータベースへの接続
        String fullUrl = url + "?createDatabaseIfNotExist=true";
        return DriverManager.getConnection(fullUrl, user, password);
    }


    private Connection getSQLiteConnection() throws SQLException {
        // プラグインのデータフォルダ内にデータベースファイルを保存する
        String fullUrl = url + "?createDatabaseIfNotExist=true";
        String sqliteUrl = "jdbc:sqlite:" + Digger.getInstance().getDataFolder().getAbsolutePath() + "/Database.db";
        return DriverManager.getConnection(sqliteUrl);
    }





    public void savePlayerData(Map<UUID, Digger.PlayerData> blockCount, List<Location> placedBlocks, Map<Location, UUID> placedBlocksWithUUID) {
        // プレイヤーデータの保存
        String playerDataQuery = "INSERT INTO digger_player_data (UUID, PlayerName, BlocksMined) VALUES (?, ?, ?) ON DUPLICATE KEY UPDATE BlocksMined = ?;";
        for (Map.Entry<UUID, Digger.PlayerData> entry : blockCount.entrySet()) {
            UUID playerId = entry.getKey();
            Digger.PlayerData playerData = entry.getValue();

            try (Connection conn = getConnection();
                 PreparedStatement stmt = conn.prepareStatement(playerDataQuery)) {

                stmt.setString(1, playerId.toString());
                stmt.setString(2, playerData.getPlayerName());
                stmt.setInt(3, playerData.getBlocksMined());
                stmt.setInt(4, playerData.getBlocksMined());

                stmt.executeUpdate();
            } catch (SQLException e) {
                e.printStackTrace();
            }
        }

        // placedBlocks の保存
        String placedBlocksQuery = "INSERT INTO digger_placed_blocks (UUID, World, X, Y, Z) VALUES (?, ?, ?, ?, ?);";

        for (Location loc : placedBlocks) {
            try (Connection conn = getConnection();
                 PreparedStatement stmt = conn.prepareStatement(placedBlocksQuery)) {

                UUID playerId = placedBlocksWithUUID.get(loc); // ここでUUIDを取得

                stmt.setString(1, playerId.toString());
                stmt.setString(2, loc.getWorld().getName());
                stmt.setInt(3, loc.getBlockX());
                stmt.setInt(4, loc.getBlockY());
                stmt.setInt(5, loc.getBlockZ());

                stmt.executeUpdate();
            } catch (SQLException e) {
                e.printStackTrace();
            }
        }
    }

    public boolean connect() {
        try (Connection conn = getConnection()) {
            return true;
        } catch (SQLException e) {
            e.printStackTrace();
            return false;
        }
    }

    public boolean isConnected() {
        try (Connection conn = getConnection()) {
            return conn != null && !conn.isClosed();
        } catch (SQLException e) {
            e.printStackTrace();
            return false;
        }
    }

    public Map<UUID, Digger.PlayerData> loadData() {
        Map<UUID, Digger.PlayerData> dataMap = new HashMap<>();
        String query = "SELECT * FROM digger_player_data;"; // SQLクエリ

        try (Connection conn = getConnection();
             PreparedStatement stmt = conn.prepareStatement(query);
             ResultSet rs = stmt.executeQuery()) {

            while (rs.next()) {
                UUID uuid = UUID.fromString(rs.getString("UUID"));
                String playerName = rs.getString("PlayerName");
                int blocksMined = rs.getInt("BlocksMined");

                dataMap.put(uuid, new Digger.PlayerData(playerName, blocksMined));
            }
        } catch (SQLException e) {
            e.printStackTrace(); // エラーログを出力
        }
        return dataMap;
    }


    public void savePlacedBlock(UUID playerName, Location loc) {
        String insertQuery = "INSERT INTO digger_placed_blocks (World, X, Y, Z) VALUES (?, ?, ?, ?);";

        try (Connection conn = getSQLiteConnection(); // SQLiteデータベースへの接続を取得
             PreparedStatement stmt = conn.prepareStatement(insertQuery)) {

            stmt.setString(1, loc.getWorld().getName());
            stmt.setInt(2, loc.getBlockX());
            stmt.setInt(3, loc.getBlockY());
            stmt.setInt(4, loc.getBlockZ());

            stmt.executeUpdate();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

}

File: src/main/java/net/hotamachisubaru/digger/mysql/MySQLDatabase.kt
package net.hotamachisubaru.digger.mysql

import net.hotamachisubaru.digger.Digger
import org.bukkit.Location
import java.io.FileInputStream
import java.io.IOException
import java.sql.Connection
import java.sql.DriverManager
import java.sql.SQLException
import java.util.Properties
import java.util.UUID
import kotlin.collections.iterator
import kotlin.use

public class MySQLDatabase(prop: Properties) {
    // データベース接続のパラメータ
    private var url = ""
    private var user = ""
    private var password = ""

    // コンストラクタでMySQLDatabaseを初期化
    init {
        try {
            // 設定ファイルからプロパティを読み込む
            prop.load(FileInputStream(Digger.getInstance().dataFolder.absolutePath + CONFIG_FILE_PATH))
            this.url = prop.getProperty("db.url", "")
            this.user = prop.getProperty("db.user", "")
            this.password = prop.getProperty("db.password", "")
            initializeDatabase()
        } catch (e: IOException) {
            // 設定ファイルが読み込めなかった場合にエラーログを出力
            Digger.getInstance().logger.severe("MySQLデータベースに接続できませんでした: ${e.message}")
        }
    }

    // データベースを初期化し、必要なテーブルを作成
    private fun initializeDatabase() {
        try {
            connection.use { conn ->
                conn.createStatement().use { stmt ->
                    // プレイヤーデータテーブルの作成（DiamondMined）
                    stmt.execute(PLAYER_DATA_TABLE)
                    // ダイヤ設置テーブルの作成（diamond_mined）
                    stmt.execute(DIAMOND_MINED_TABLE)
                }
            }
        } catch (e: SQLException) {
            // データベースの初期化が失敗した場合にエラーログを出力
            Digger.getInstance().logger.severe("MySQLデータベース初期化エラー: ${e.message}")
        }
    }

    @get:Throws(SQLException::class)
    private val connection: Connection
        // MySQLデータベースへの接続を確立
        get() {
            // データベースが存在しない場合は作成する
            val fullUrl = "$url?createDatabaseIfNotExist=true"
            return DriverManager.getConnection(fullUrl, user, password)
        }

    /**
     * プレイヤーデータとダイヤ設置データを保存
     */
    fun saveData(
        diamondCount: Map<UUID, Digger.PlayerData>,
        diamondLocations: List<Location>,
        diamondLocationsWithUUID: Map<Location, UUID>
    ) {
        saveDiamondCountEntries(diamondCount)
        saveDiamondLocationEntries(diamondLocations, diamondLocationsWithUUID)
    }

    // ダイヤ採掘数データを保存
    private fun saveDiamondCountEntries(diamondCount: Map<UUID, Digger.PlayerData>) {
        val query = """
            INSERT INTO player_data (UUID, PlayerName, DiamondMined)
            VALUES (?, ?, ?)
            ON DUPLICATE KEY UPDATE DiamondMined = VALUES(DiamondMined), PlayerName = VALUES(PlayerName);
        """.trimIndent()
        try {
            connection.use { conn ->
                conn.prepareStatement(query).use { stmt ->
                    for ((uuid, data) in diamondCount) {
                        stmt.setString(1, uuid.toString())
                        stmt.setString(2, data.playerName)
                        stmt.setInt(3, data.getDiamondMined())
                        stmt.executeUpdate()
                    }
                }
            }
        } catch (e: SQLException) {
            Digger.getInstance().logger.severe("ダイヤ採掘数の保存に失敗しました: ${e.message}")
        }
    }

    // ダイヤ設置データを保存
    private fun saveDiamondLocationEntries(
        locations: List<Location>,
        withUUID: Map<Location, UUID>
    ) {
        val query = "INSERT INTO diamond_mined (UUID, World, X, Y, Z) VALUES (?, ?, ?, ?, ?);"
        try {
            connection.use { conn ->
                conn.prepareStatement(query).use { stmt ->
                    for (loc in locations) {
                        val uuid = withUUID[loc] ?: continue
                        stmt.setString(1, uuid.toString())
                        stmt.setString(2, loc.world.name)
                        stmt.setInt(3, loc.blockX)
                        stmt.setInt(4, loc.blockY)
                        stmt.setInt(5, loc.blockZ)
                        stmt.executeUpdate()
                    }
                }
            }
        } catch (e: SQLException) {
            Digger.getInstance().logger.severe("ダイヤ設置データの保存に失敗しました: ${e.message}")
        }
    }

    /**
     * データベースからプレイヤーデータを読み込む
     */
    fun loadData(): Map<UUID, Digger.PlayerData> {
        val result = mutableMapOf<UUID, Digger.PlayerData>()
        val query = "SELECT * FROM player_data;"
        try {
            connection.use { conn ->
                conn.prepareStatement(query).use { stmt ->
                    stmt.executeQuery().use { rs ->
                        while (rs.next()) {
                            val uuid = UUID.fromString(rs.getString("UUID"))
                            val name = rs.getString("PlayerName")
                            val count = rs.getInt("DiamondMined")
                            result[uuid] = Digger.PlayerData(name, count)
                        }
                    }
                }
            }
        } catch (e: SQLException) {
            Digger.getInstance().logger.severe("データ読み込みに失敗しました: ${e.message}")
        }
        return result
    }

    // データベース接続が有効か確認
    fun isConnected(): Boolean {
        return try {
            connection.use { conn -> conn.isValid(5) } // タイムアウト2秒
        } catch (e: SQLException) {
            Digger.getInstance().logger.severe("データベース接続の確認に失敗しました: ${e.message}")
            false
        }
    }

    companion object {
        private const val CONFIG_FILE_PATH = "/config.properties"

        // テーブル定義
        private const val PLAYER_DATA_TABLE = (
                "CREATE TABLE IF NOT EXISTS player_data (" +
                        "UUID CHAR(36) PRIMARY KEY, " +
                        "PlayerName VARCHAR(255), " +
                        "DiamondMined INT" +
                        ");"
                )
        private const val DIAMOND_MINED_TABLE = (
                "CREATE TABLE IF NOT EXISTS diamond_mined (" +
                        "UUID CHAR(36), " +
                        "World VARCHAR(255), " +
                        "X INT, " +
                        "Y INT, " +
                        "Z INT, " +
                        "FOREIGN KEY(UUID) REFERENCES player_data(UUID)" +
                        ");"
                )
    }
}